File: python/bKap5_5_InSamplePCA_GZW.py
"""
bKap5_5_InSamplePCA_GZW.py
Python port of bKap5_5_InSamplePCA_GZW.m.

PCA dimensionality reduction of the Liedtke macro panel (in-sample). The macro
predictors are reduced to a few principal components and the inflation target is
regressed on those components (PCR).

Run from the repo root:  python python/bKap5_5_InSamplePCA_GZW.py
"""

import os
import numpy as np
import pandas as pd
import scipy.io as sio

import models

# =========================================================================
#  Settings  (mirror the MATLAB header)
# =========================================================================
DATA_PATH    = "./DATA/Liedtke/US/aggregated.csv"
RESULTS_PATH = "./RESULTS/GWZ/"

iTimeLag    = 1     # additional predictive lag (predictors already lag-aligned)
iNumComp    = 3     # number of principal components to retain
iTransformX = 2     # 0 = none, 1 = centre, 2 = z-standardise
SHOW_PLOT   = False  # set True to draw the explained-variance bar chart

# =========================================================================
#  Load data
# =========================================================================
tData = pd.read_csv(DATA_PATH)
cAllNames = list(tData.columns)
dtDates = pd.to_datetime(tData.iloc[:, 0])
mX = tData.iloc[:, 2:].to_numpy(float)     # macro predictors
vY = tData.iloc[:, 1].to_numpy(float)      # inflation target
cXnamesM = cAllNames[2:]

iNumObs, iNumPredictors = mX.shape

# Lag the predictors so they are known one period ahead of the target
mXlag = np.full_like(mX, np.nan)
mXlag[iTimeLag:, :] = mX[:-iTimeLag, :]

# Remove missing values row-wise (PCA does not allow missing values)
lIsNaN = np.isnan(vY) | np.isnan(mXlag).any(axis=1)
keep = ~lIsNaN
vY = vY[keep]
mXlag = mXlag[keep, :]
dtDates = dtDates[keep].reset_index(drop=True)

iNumObs = len(vY)
iNumComp = min(iNumComp, iNumPredictors)

# =========================================================================
#  PCA dimensionality reduction
# =========================================================================
# Full PCA to obtain explained variance for all factors
rModelFull = models.est_pca(mXlag, num_comp=iNumPredictors, transform_x=iTransformX)
vExplVarPerFactor = np.diff(np.concatenate([[0.0], rModelFull["expl_var"]])) * 100

if SHOW_PLOT:
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt
    plt.figure()
    plt.bar(np.arange(1, iNumPredictors + 1), vExplVarPerFactor)
    plt.xlabel("Factor")
    plt.ylabel("Explained Variance (%)")
    plt.title("Explained Variance by Factor")
    plt.box(False)
    plt.savefig(os.path.join(RESULTS_PATH, "InSamplePCA_ExplVar_py.png"))

# Estimate PCA keeping iNumComp principal components
rModel = models.est_pca(mXlag, num_comp=iNumComp, transform_x=iTransformX)
mScores = rModel["scores"]

# =========================================================================
#  In-sample analysis: regress inflation on the retained PCs (with intercept)
# =========================================================================
reg = models.ols_regress(vY, mScores)
vBeta = reg["beta"]
vBetaT = reg["t"]
dR2 = reg["R2"]
vYhat = reg["yhat"]

# =========================================================================
#  Report
# =========================================================================
print(f"In-sample PCA regression: inflation ~ {iNumComp} principal components")
print(f"  Variance explained by {iNumComp} PCs : {rModel['expl_var'][-1] * 100:6.2f}%")
print(f"  Intercept     : {vBeta[0]:8.4f} (t = {vBetaT[0]:6.2f})")
for k in range(iNumComp):
    print(f"  PC{k + 1:<2d} coef     : {vBeta[k + 1]:8.4f} (t = {vBetaT[k + 1]:6.2f})")
print(f"  R2            : {dR2:8.4f}")

# =========================================================================
#  Save results (Python copy; does not overwrite the MATLAB .mat)
# =========================================================================
os.makedirs(RESULTS_PATH, exist_ok=True)
sio.savemat(os.path.join(RESULTS_PATH, "InSamplePCAResults_py.mat"), {
    "vBeta": vBeta.reshape(-1, 1),
    "vBetaT": vBetaT.reshape(-1, 1),
    "dR2": dR2,
    "vYhat": vYhat.reshape(-1, 1),
    "rModel": {
        "vMeanX": rModel["mean_x"], "vStdX": rModel["std_x"],
        "iNumComp": rModel["num_comp"], "mEigVec": rModel["eigvec"],
        "vEigVal": rModel["eigval"].reshape(-1, 1), "mScores": rModel["scores"],
        "vExplVar": rModel["expl_var"].reshape(-1, 1),
    },
})

File: bKap5_5_InSamplePCA_GZW.m
% Script for applying PCA dimensionality reduction to the Liedtke macro panel
% (in-sample). The macro predictors are reduced to a few principal components
% and the inflation target is regressed on those components (PCR).

% Clear console
clear; clc; close all;

% Set paths
sOldPath = path;
sCountry = fCfg('COUNTRY', 'US');
sDataPath = ['./DATA/Liedtke/', sCountry, '/'];
sResultsPath = './RESULTS/GWZ/';
addpath('./Utils/');

%% Settings
iTimeLag    = 1;     % additional predictive lag (predictors are already
                     % reporting-lag aligned in DATA/Liedtke/aggregate.py)
iNumComp    = 3;     % number of principal components to retain
iTransformX = 2;     % PCA preprocessing: 0 = none, 1 = centre, 2 = z-standardise.
                     % The predictors live on very different scales (rate diffs
                     % vs percentage changes), so standardisation is used.

%% Load data
% Macro panel produced by DATA/Liedtke/aggregate.py.
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (already lag-aligned in Python).
tData     = readtable([sDataPath, 'aggregated.csv']);
cAllNames = tData.Properties.VariableNames;

% Parse dates (first column)
dtDates = tData{:,1};
if ~isdatetime(dtDates)
    dtDates = datetime(dtDates);
end

% Target (col 2) and predictors (col 3..end)
mX          = tData{:,3:end};          % macro predictors
vY          = tData{:,2};              % inflation target
cXnamesM    = cAllNames(3:end)';       % predictor names as an Nx1 cell

%% Data preprocessing
% Determine dimensions
[iNumObs, iNumAssets]       = size(vY);
[iNumObsP, iNumPredictors]  = size(mX);

% Lag the predictors so that they are known one period ahead of the target
mXlag = [NaN(iTimeLag, iNumPredictors); mX(1:end-iTimeLag,:)];

% Remove missing values row-wise (PCA does not allow missing values)
lIsNaN          = isnan(vY) | any(isnan(mXlag),2);
vY(lIsNaN)      = [];
mXlag(lIsNaN,:) = [];
dtDates(lIsNaN) = [];

% Update number of observations after NaN removal
iNumObs  = size(vY,1);
iNumComp = min(iNumComp, iNumPredictors);

%% PCA dimensionality reduction
% Run full PCA to obtain explained variance for all factors
rModelFull = fEstPCA(mXlag, "iNumComp", iNumPredictors, "iTransformX", iTransformX);

% Plot explained variance per factor (individual, not cumulative)
vExplVarPerFactor = diff([0; rModelFull.vExplVar]) * 100;
figure;
bar(1:iNumPredictors, vExplVarPerFactor);
xlabel('Factor');
ylabel('Explained Variance (%)');
title('Explained Variance by Factor');
box off;

% Estimate PCA keeping iNumComp principal components
rModel  = fEstPCA(mXlag, "iNumComp", iNumComp, "iTransformX", iTransformX);
mScores = rModel.mScores;

%% In-Sample analysis
% Regression with intercept using the retained principal components
rResults = regstats2(vY, mScores, 'linear', {'tstat','rsquare'});
vBeta    = rResults.tstat.beta;   % coefficient estimates
vBetaT   = rResults.tstat.t;      % t-Statistics
dR2      = rResults.rsquare;      % R2

% In-sample prediction
vYhat = [ones(iNumObs,1), mScores] * vBeta;

% === Report
fprintf('In-sample PCA regression: inflation ~ %d principal components\n', iNumComp);
fprintf('  Variance explained by %d PCs : %6.2f%%\n', iNumComp, rModel.vExplVar(end)*100);
fprintf('  Intercept     : %8.4f (t = %6.2f)\n', vBeta(1), vBetaT(1));
for k = 1:iNumComp
    fprintf('  PC%-2d coef     : %8.4f (t = %6.2f)\n', k, vBeta(k+1), vBetaT(k+1));
end
fprintf('  R2            : %8.4f\n', dR2);

% === Save results
% Structured output: <GWZ>/PCA/<country>/insample/<options>/
sCountry = fCountryFromPath(sDataPath);
sOutDir  = fResultDir(sResultsPath, 'PCA', sCountry, 'insample', ...
    sprintf('comp%d_lag%d', iNumComp, iTimeLag));
exportgraphics(gcf, fullfile(sOutDir, 'chart.png'), 'Resolution', 150);
tRes = cell2table({vBeta(1), dR2}, 'VariableNames', {'Intercept', 'R2'});
writetable(tRes, fullfile(sOutDir, 'results.csv'));
sFilename = fullfile(sOutDir, 'results.mat');
save(sFilename, 'vBeta', 'vBetaT', 'dR2', 'vYhat', 'rModel');

% Restore path
path(sOldPath);
